content/advent-2015/building-static-linuxes.md

```diff
@@ -0,0 +1,94 @@
++++
+author = ["Ian Chiles (fortytw2)"]
+date = "2015-12-24T00:00:00-08:00"
+series = ["Advent 2015"]
+title = "Tiny Linux OSes with Go"
+
++++
+
+> Small disclaimer: This is much more _fun_ than it is _useful_
+
+For a while now, we've been seeing new "cloud" OSes crop up, like CoreOS and
+RancherOS. These two both simply marry docker+linux to create a magic docker
+runtime.
+
+You may be familiar with "micro" Docker images built on top of Alpine Linux /
+BusyBox base images. These "images" general weigh in around 10-30mb. But that's
+before including the ~300mb OS + actual Docker daemon that they have to run on -
+seems awfully bloated to me.
+
+So, we need to ask oursevles, "who really needs docker when we can have an
+entire OS that is purpose built to run your application, and nothing else? Short answer,
+most people; fun answer, no one.
+
+So let's get started. Make sure you have a copy of `QEMU` installed locally,
+as we're going to _cheat_ and use its ability to boot a raw kernel bzImage
+(the compressed, bootable kernel image) to avoid having to set up a real
+bootloader (not too hard, but can be a real pain).
+
+So, to get in the right mindset, we're going to use the excellent `termboy-go`
+as our test application, an incredibly cool Gameboy Color emulator that runs
+exclusively in the Linux console. In the end, we'll have a "single purpose OS"
+that just runs a Gameboy emulator.
+
+# Linux
+
+First off, we're going to need a Linux kernel, so grab mainline from kernel.org,
+run `make menuconfig` and then `make bzImage`. The default config should work,
+just ensure that EXT4 is compiled *not* as a module.
+
+<script type="text/javascript" src="https://asciinema.org/a/4adm32wk0t626mijvkexuvtmk.js" id="asciicast-4adm32wk0t626mijvkexuvtmk" async></script>
+
+To boot our new kernel and make sure everything works, just use QEMU. You should
+get a `Kernel panic - not syncing: VFS: Unable to mount root fs on unknown-block(0,0)`
+which makes complete sense, as we haven't added any sort of root FS.
+
+<script type="text/javascript" src="https://asciinema.org/a/12i116g0kltwwxn2c8jwmsq81.js" id="asciicast-12i116g0kltwwxn2c8jwmsq81" async></script>
+
+Using a tool within `libguestfs`, `virt-make-fs` we can create a QEMU compatible
+`.qcow2` from a directory (mine is a directory named `null`) and boot with it.
+
+<script type="text/javascript" src="https://asciinema.org/a/bfk43pq02kdq0hq4das6q70f3.js" id="asciicast-bfk43pq02kdq0hq4das6q70f3" async></script>
+
+Looking good, but we got another kernel panic - `Kernel panic - not syncing: No
+working init found. Try passing init= option to kernel. See Linux
+Documentation/init.txt for guidance.`. Luckily, this one's easy to solve, as,
+funnily enough, any binary can be used as the kernel `init` - Go just so happens
+to make wonderful, static binaries.
+
+# Termboy-go
```

Comment:
There are no links to termboy-go in the post. It may be worth putting one in so someone can click through to it.